File: android/app/src/main/java/info/bagen/dwebbrowser/microService/sys/nativeui/safeArea/SafeAreaController.kt
package info.bagen.dwebbrowser.microService.sys.nativeui.safeArea


import androidx.activity.ComponentActivity
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.displayCutout
import androidx.compose.foundation.layout.exclude
import androidx.compose.foundation.layout.union
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import info.bagen.dwebbrowser.microService.helper.InsetsJson
import info.bagen.dwebbrowser.microService.helper.toJsonAble
import info.bagen.dwebbrowser.microService.sys.nativeui.NativeUiController
import info.bagen.dwebbrowser.microService.sys.nativeui.base.InsetsController
import info.bagen.dwebbrowser.microService.sys.nativeui.helper.debugNativeUi
import info.bagen.dwebbrowser.microService.sys.nativeui.helper.plus

/**
 * 安全区域，被 设备的顶部流海、状态栏、导航栏等原生UI所影响后，分割出inner、outer两个部分
 */
class SafeAreaController(
    activity: ComponentActivity, nativeUiController: NativeUiController
) : InsetsController(activity, nativeUiController) {
    /**
     * 刘海屏
     */
    val cutoutInsetsState = mutableStateOf(WindowInsets(0))


    /**
     * 外部区域
     */
    val outerAreaInsetsState = mutableStateOf(WindowInsets(0))


    @Composable
    override fun effect(): SafeAreaController {
        cutoutInsetsState.value = WindowInsets.Companion.displayCutout

        val statusBar = nativeUiController.statusBar
        val virtualKeyboard = nativeUiController.virtualKeyboard
        val navigationBar = nativeUiController.navigationBar
        val safeArea = this

        observer.stateChanges.let {
            val isSafeAreaOverlay by it.rememberByState(safeArea.overlayState)
            val safeAreaCutoutInsets by it.rememberByState(safeArea.cutoutInsetsState)

            val isStatusBarOverlay by it.rememberByState(statusBar.overlayState)
            val statusBarsInsets by it.rememberByState(statusBar.insetsState)

            val isVirtualKeyboardOverlay by it.rememberByState(virtualKeyboard.overlayState)
            val imeInsets by it.rememberByState(virtualKeyboard.insetsState)

            val isNavigationBarOverlay by it.rememberByState(navigationBar.overlayState)
            val navigationBarsInsets by it.rememberByState(navigationBar.insetsState)

            it.effectChange {
                var RES_safeAreaInsets = WindowInsets(0)
                var RES_outerAreaArea = WindowInsets(0)

                /// 顶部，顶部有状态栏和刘海区域
                val topInsets = statusBarsInsets.union(safeAreaCutoutInsets)
                if (isStatusBarOverlay && isSafeAreaOverlay) {
                    // 都覆盖，那么就写入safeArea，outerArea不需要调整
                    RES_safeAreaInsets += topInsets
                } else if (isStatusBarOverlay && !isSafeAreaOverlay) {
                    // outerArea写入刘海区域，safeArea只写剩余的
                    RES_outerAreaArea += safeAreaCutoutInsets
                    RES_safeAreaInsets += topInsets.exclude(safeAreaCutoutInsets)
                } else if (!isStatusBarOverlay && isSafeAreaOverlay) {
                    // outerArea写入状态栏，safeArea只写剩余的
                    RES_outerAreaArea += statusBarsInsets
                    RES_safeAreaInsets += topInsets.exclude(statusBarsInsets)
                } else {
                    // 都不覆盖，全部写入 outerArea
                    RES_outerAreaArea += topInsets
                }
                /// 底部，底部有导航栏和虚拟键盘
                val bottomInsets = navigationBarsInsets.union(imeInsets)
                if (isVirtualKeyboardOverlay && isNavigationBarOverlay) {
                    // 都覆盖，那么就写入safeArea，outerArea不需要调整
                    RES_safeAreaInsets += bottomInsets
                } else if (isVirtualKeyboardOverlay && !isNavigationBarOverlay) {
                    // outerArea写入导航栏，safeArea只写剩余的
                    RES_outerAreaArea += navigationBarsInsets
                    RES_safeAreaInsets += bottomInsets.exclude(navigationBarsInsets)
                } else if (!isVirtualKeyboardOverlay && isNavigationBarOverlay) {
                    // outerArea写入虚拟键盘，safeArea只写剩余的
                    RES_outerAreaArea += imeInsets
                    RES_safeAreaInsets += bottomInsets.exclude(imeInsets)
                } else if (!isVirtualKeyboardOverlay && !isNavigationBarOverlay) {
                    // 都不覆盖，全部写入 outerArea
                    RES_outerAreaArea += bottomInsets
                }

                insetsState.value = RES_safeAreaInsets
                outerAreaInsetsState.value = RES_outerAreaArea
                debugNativeUi("SafeArea", "CHANGED")
                observer.notifyObserver()
            }
        }
        return this
    }


    data class SafeAreaState(
      val cutoutInsets: InsetsJson,
      val outerInsets: InsetsJson,
      override val overlay: Boolean,
      override val insets: InsetsJson,
    ) : InsetsState

    override fun toJsonAble() = SafeAreaState(
        cutoutInsets = cutoutInsetsState.value.toJsonAble(),
        overlay = overlayState.value,
        outerInsets = outerAreaInsetsState.value.toJsonAble(),
        insets = insetsState.value.toJsonAble(),
    )
}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/NativeUI/SafeArea/SafeAreaController.cs
using UIKit;
using CoreGraphics;
using System.Text.Json;
using System.Text.Json.Serialization;
using DwebBrowser.MicroService.Sys.NativeUI;
using DwebBrowser.MicroService.Sys.Mwebview;
using DwebBrowser.Platforms.iOS.MicroModule.NativeUI.Base;
using DwebBrowser.Helper;

namespace DwebBrowser.Platforms.iOS.MicroModule.NativeUI.SafeArea;

public class SafeAreaController : AreaController, IToJsonAble
{
    static Debugger Console = new("SafeAreaController");

    // Notch 刘海 or Dynamic Island 灵动岛
    public State<AreaJson> CutoutInsets { get; init; }

    // 外部区域
    public State<AreaJson> OuterAreaInsets { get; init; }

    public readonly State<SafeAreaState> Observer;
    public StateObservable<SafeAreaState> StateObserver { get; init; }

    public NativeUiController NativeUiController { get; init; }

    public SafeAreaController(
        MultiWebViewController mwebviewController,
        NativeUiController nativeUiController) : base(
        overlayState: new(true),
        areaState: new(CGRect.Empty.ToAreaJson()))
    {
        CutoutInsets = new(new AreaJson(UIApplication.SharedApplication.KeyWindow.SafeAreaInsets.Top, 0, 0, 0));
        var webviewFrame = mwebviewController.WebviewFrame.Get();
        OuterAreaInsets = new(new AreaJson(
            webviewFrame.Top.Value,
            0,
            0,
            mwebviewController.webviewContainer.Frame.GetMaxY().Value
            - webviewFrame.Bottom.Value));
        Observer = new State<SafeAreaState>(() => GetState());
        StateObserver = new(() => ToJson());
        NativeUiController = nativeUiController;

        var statusBar = NativeUiController.StatusBar;
        var navigationBar = NativeUiController.NavigationBar;
        var virtualKeyboard = NativeUiController.VirtualKeyboard;
        var safeArea = this;

        Observer.OnChange += async (value, oldValue, _) =>
        {
            await MainThread.InvokeOnMainThreadAsync(() =>
            {
                var isSafeAreaOverlay = value.Overlay;
                var safeAreaCutoutInsets = safeArea.CutoutInsets.Get();

                var statusBarState = statusBar.GetState();
                var isStatusBarOverlay = statusBarState.Overlay;
                var statusBarInsets = statusBarState.Area;

                var navigationBarState = navigationBar.GetState();
                var isNavigationBarOverlay = navigationBarState.Overlay;
                var navigationBarInsets = navigationBarState.Area;

                var virtualKeyboardState = virtualKeyboard.GetState();
                var isVirtualKeyboardOverlay = virtualKeyboardState.Overlay;
                var virtualKeyboardInsets = virtualKeyboardState.Area;

                var RES_safeAreaInsets = AreaJson.Empty;
                var RES_outerAreaInsets = AreaJson.Empty;

                /// 顶部有状态栏、刘海或灵动岛区域
                var topInsets = statusBarInsets.Union(safeAreaCutoutInsets);

                if (isStatusBarOverlay && isSafeAreaOverlay)
                {
                    // 都覆盖，那么就写入safeArea，outerArea不需要调整
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(topInsets);
                }
                else if (isStatusBarOverlay && !isSafeAreaOverlay)
                {
                    // outerArea写入刘海区域，safeArea只写剩余的
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(safeAreaCutoutInsets);
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(topInsets.Exclude(safeAreaCutoutInsets));
                }
                else if (!isStatusBarOverlay && isSafeAreaOverlay)
                {
                    // outerArea写入状态栏，safeArea只写剩余的
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(statusBarInsets);
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(topInsets.Exclude(statusBarInsets));
                }
                else
                {
                    // 都不覆盖，全部写入 outerArea
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(topInsets);
                }

                /// 底部，底部有导航栏和虚拟键盘
                var bottomInsets = navigationBarInsets.Union(virtualKeyboardInsets);

                if (isVirtualKeyboardOverlay && isNavigationBarOverlay)
                {
                    // 都覆盖，那么就写入safeArea，outerArea不需要调整
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(bottomInsets);
                }
                else if (isVirtualKeyboardOverlay && !isNavigationBarOverlay)
                {
                    // outerArea写入导航栏，safeArea只写剩余的
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(navigationBarInsets);
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(bottomInsets.Exclude(navigationBarInsets));
                }
                else if (!isVirtualKeyboardOverlay && isNavigationBarOverlay)
                {
                    // outerArea写入虚拟键盘，safeArea只写剩余的
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(virtualKeyboardInsets);
                    RES_safeAreaInsets = RES_safeAreaInsets.Add(bottomInsets.Exclude(virtualKeyboardInsets));
                }
                else
                {
                    // 都不覆盖，全部写入 outerArea
                    RES_outerAreaInsets = RES_outerAreaInsets.Add(bottomInsets);
                }

                AreaState.Set(RES_safeAreaInsets);
                OuterAreaInsets.Set(RES_outerAreaInsets);
                mwebviewController.WebviewFrame.Set(new CGRect(
                    new nfloat(RES_outerAreaInsets.left),
                    new nfloat(RES_outerAreaInsets.top),
                    mwebviewController.webviewContainer.Frame.Width,
                    new nfloat(
                        mwebviewController.webviewContainer.Frame.GetMaxY().Value
                        - RES_outerAreaInsets.bottom
                        - RES_outerAreaInsets.top)));
            });

            await StateObserver.EmitAsync();
        };
    }

    public SafeAreaState GetState()
    {
        return new SafeAreaState
        {
            CutoutInsets = CutoutInsets.Get(),
            OuterAreaInsets = OuterAreaInsets.Get(),
            Overlay = OverlayState.Get(),
            Area = AreaState.Get()
        };
    }


    public string ToJson()
    {
        return JsonSerializer.Serialize(GetState());
    }
}

public class SafeAreaState : AreaState
{
    [JsonPropertyName("cutoutInsets")]
    public AreaJson CutoutInsets { get; set; }
    [JsonPropertyName("outerInsets")]
    public AreaJson OuterAreaInsets { get; set; }
}